qiskit/circuit/parameterexpression.py

```diff
@@ -255,6 +272,9 @@ def __int__(self):
                             'cannot be cast to an int.'.format(self.parameters))
         return int(self._symbol_expr)
 
+    def __hash__(self):
+        return hash((tuple(sorted(self.parameters, key=lambda p: p.name)), str(self)))
```

Comment:
Maybe you can avoid the ordering issue with,
`hash((frozenset(self._parameter_symbols), self._symbol_expr))`